File: services/backend/core/cloudflare.py
"""Async HTTP client for Cloudflare AI Search (managed RAG)."""

from __future__ import annotations

import logging
from typing import Any
from urllib.parse import quote

import httpx

from core.config import settings

log = logging.getLogger(__name__)

API_BASE = "https://api.cloudflare.com/client/v4"
MAX_FILE_BYTES = 4 * 1024 * 1024
_SEARCH_MAX_RESULTS = 50


class CloudflareAISearchError(RuntimeError):
    """Raised when the Cloudflare AI Search API rejects a request."""


def is_configured() -> bool:
    return bool(
        settings.cloudflare_account_id
        and settings.cloudflare_api_token
        and settings.cloudflare_ai_search_instance
    )


def item_key(collection: str, document_id: str, filename: str) -> str:
    """Object key used as the folder prefix for per-professor retrieval."""
    safe_name = filename.replace("\\", "/").split("/")[-1] or "document"
    return f"{collection}/{document_id}/{safe_name}"


def parse_item_key(key: str) -> tuple[str, str, str]:
    """Return (collection, document_id, filename) from an item key."""
    parts = key.split("/")
    if len(parts) >= 3:
        return parts[0], parts[1], parts[-1]
    if len(parts) == 2:
        return parts[0], "", parts[1]
    return "", "", key


def folder_starts_with_filter(prefix: str) -> dict[str, Any]:
    """Vectorize-style 'starts with' filter for a folder prefix."""
    if not prefix.endswith("/"):
        prefix = f"{prefix}/"
    return {"folder": {"$gte": prefix, "$lt": f"{prefix[:-1]}0"}}


def _instance_url(path: str = "") -> str:
    base = (
        f"{API_BASE}/accounts/{settings.cloudflare_account_id}"
        f"/ai-search/instances/{settings.cloudflare_ai_search_instance}"
    )
    return f"{base}/{path.lstrip('/')}" if path else base


def _auth_headers() -> dict[str, str]:
    return {"Authorization": f"Bearer {settings.cloudflare_api_token}"}


def _unwrap(payload: Any) -> Any:
    if isinstance(payload, dict) and "result" in payload:
        if payload.get("success") is False:
            errors = payload.get("errors") or payload.get("error") or payload
            raise CloudflareAISearchError(f"Cloudflare AI Search error: {errors}")
        return payload["result"]
    return payload


def _require_configured() -> None:
    if not is_configured():
        raise CloudflareAISearchError(
            "Cloudflare AI Search is not configured. Set CLOUDFLARE_ACCOUNT_ID, "
            "CLOUDFLARE_API_TOKEN, and CLOUDFLARE_AI_SEARCH_INSTANCE."
        )


async def get_instance() -> dict[str, Any]:
    """Fetch the configured AI Search instance (used as a health probe)."""
    _require_configured()
    async with httpx.AsyncClient(timeout=settings.cloudflare_timeout_seconds) as client:
        response = await client.get(_instance_url(), headers=_auth_headers())
        response.raise_for_status()
        result = _unwrap(response.json())
    if not isinstance(result, dict):
        return {"id": settings.cloudflare_ai_search_instance}
    return result


async def upload_item(
    key: str,
    content: bytes,
    content_type: str,
    *,
    wait_for_completion: bool = True,
) -> dict[str, Any]:
    """Upload a file into the instance's built-in storage and index it."""
    _require_configured()
    if len(content) > MAX_FILE_BYTES:
        raise CloudflareAISearchError(
            f"FILE_TOO_LARGE: Cloudflare AI Search rejects files over 4 MB "
            f"({len(content)} bytes)"
        )
    data = {"wait_for_completion": "true" if wait_for_completion else "false"}
    files = {"file": (key, content, content_type)}
    async with httpx.AsyncClient(timeout=settings.cloudflare_timeout_seconds) as client:
        response = await client.post(
            _instance_url("items"),
            headers=_auth_headers(),
            data=data,
            files=files,
        )
        response.raise_for_status()
        result = _unwrap(response.json())
    if not isinstance(result, dict):
        return {"key": key, "status": "queued"}
    return result


async def get_item_by_key(key: str) -> dict[str, Any] | None:
    _require_configured()
    async with httpx.AsyncClient(timeout=settings.cloudflare_timeout_seconds) as client:
        response = await client.get(
            _instance_url("items"),
            headers=_auth_headers(),
            params={"key": key, "source": "builtin"},
        )
        response.raise_for_status()
        result = _unwrap(response.json())
    items = _as_item_list(result)
    for item in items:
        if item.get("key") == key:
            return item
    return items[0] if items else None


async def list_items() -> list[dict[str, Any]]:
    _require_configured()
    items: list[dict[str, Any]] = []
    page = 1
    async with httpx.AsyncClient(timeout=settings.cloudflare_timeout_seconds) as client:
        while True:
            response = await client.get(
                _instance_url("items"),
                headers=_auth_headers(),
                params={"page": page, "per_page": 100, "source": "builtin"},
            )
            response.raise_for_status()
            payload = response.json() if response.content else {}
            result = _unwrap(payload)
            batch = _as_item_list(result)
            items.extend(batch)
            info = payload.get("result_info") if isinstance(payload, dict) else None
            total = (info or {}).get("total_count") if isinstance(info, dict) else None
            if not batch or (total is not None and len(items) >= int(total)):
                break
            if len(batch) < 100:
                break
            page += 1
    return items


async def delete_item(item_id: str) -> None:
    _require_configured()
    encoded = quote(item_id, safe="")
    async with httpx.AsyncClient(timeout=settings.cloudflare_timeout_seconds) as client:
        response = await client.delete(
            _instance_url(f"items/{encoded}"),
            headers=_auth_headers(),
        )
        response.raise_for_status()


async def search(
    query: str,
    *,
    folder_prefix: str | None = None,
    max_num_results: int | None = None,
) -> list[dict[str, Any]]:
    """Semantic search. Returns raw chunk dicts from the API."""
    _require_configured()
    limit = max(1, min(max_num_results or settings.rag_retrieval_top_k, _SEARCH_MAX_RESULTS))
    body: dict[str, Any] = {
        "query": query or " ",
        "ai_search_options": {
            "retrieval": {
                "max_num_results": limit,
            }
        },
    }
    if folder_prefix:
        body["ai_search_options"]["retrieval"]["filters"] = folder_starts_with_filter(
            folder_prefix
        )
    async with httpx.AsyncClient(timeout=settings.cloudflare_timeout_seconds) as client:
        response = await client.post(
            _instance_url("search"),
            headers={**_auth_headers(), "Content-Type": "application/json"},
            json=body,
        )
        response.raise_for_status()
        result = _unwrap(response.json())
    if isinstance(result, dict):
        chunks = result.get("chunks") or result.get("data") or []
        return list(chunks)
    if isinstance(result, list):
        return result
    return []


def _as_item_list(result: Any) -> list[dict[str, Any]]:
    if result is None:
        return []
    if isinstance(result, list):
        return [item for item in result if isinstance(item, dict)]
    if isinstance(result, dict):
        for key in ("items", "data", "objects"):
            nested = result.get(key)
            if isinstance(nested, list):
                return [item for item in nested if isinstance(item, dict)]
        if "id" in result or "key" in result:
            return [result]
    return []

File: api/src/cloudflare.ts
import type { Env } from './types'

const API_BASE = 'https://api.cloudflare.com/client/v4'

export const MAX_FILE_BYTES = 4 * 1024 * 1024
const SEARCH_MAX_RESULTS = 50

export class CloudflareAISearchError extends Error {
  constructor(message: string) {
    super(message)
    this.name = 'CloudflareAISearchError'
  }
}

export function isConfigured(env: Env): boolean {
  return Boolean(
    env.CLOUDFLARE_ACCOUNT_ID &&
      env.CLOUDFLARE_API_TOKEN &&
      env.CLOUDFLARE_AI_SEARCH_INSTANCE,
  )
}

export function itemKey(collection: string, documentId: string, filename: string): string {
  const safeName = filename.replace(/\\/g, '/').split('/').pop() || 'document'
  return `${collection}/${documentId}/${safeName}`
}

export function parseItemKey(key: string): { collection: string; documentId: string; filename: string } {
  const parts = key.split('/')
  if (parts.length >= 3) return { collection: parts[0], documentId: parts[1], filename: parts[parts.length - 1] }
  if (parts.length === 2) return { collection: parts[0], documentId: '', filename: parts[1] }
  return { collection: '', documentId: '', filename: key }
}

export function folderStartsWithFilter(prefix: string): Record<string, Record<string, string>> {
  const normalized = prefix.endsWith('/') ? prefix : `${prefix}/`
  return { folder: { $gte: normalized, $lt: `${normalized.slice(0, -1)}0` } }
}

function instanceUrl(env: Env, path = ''): string {
  const base = `${API_BASE}/accounts/${env.CLOUDFLARE_ACCOUNT_ID}/ai-search/instances/${env.CLOUDFLARE_AI_SEARCH_INSTANCE}`
  return path ? `${base}/${path.replace(/^\//, '')}` : base
}

function authHeaders(env: Env): Record<string, string> {
  return { Authorization: `Bearer ${env.CLOUDFLARE_API_TOKEN}` }
}

function requireConfigured(env: Env): void {
  if (!isConfigured(env)) {
    throw new CloudflareAISearchError(
      'Cloudflare AI Search is not configured. Set CLOUDFLARE_ACCOUNT_ID, CLOUDFLARE_API_TOKEN, and CLOUDFLARE_AI_SEARCH_INSTANCE.',
    )
  }
}

function unwrap<T>(payload: unknown): T {
  if (payload && typeof payload === 'object' && 'result' in payload) {
    const p = payload as Record<string, unknown>
    if (p.success === false) {
      throw new CloudflareAISearchError(`Cloudflare AI Search error: ${JSON.stringify(p.errors ?? p.error ?? p)}`)
    }
    return p.result as T
  }
  return payload as T
}

export async function getInstance(env: Env, timeoutMs = 60000): Promise<Record<string, unknown>> {
  requireConfigured(env)
  try {
    const response = await fetch(instanceUrl(env), {
      headers: authHeaders(env),
      signal: AbortSignal.timeout(timeoutMs),
    })
    if (!response.ok) {
      const body = await response.text().catch(() => '')
      throw new CloudflareAISearchError(`Cloudflare instance probe failed (${response.status}): ${body.slice(0, 300)}`)
    }
    const result = unwrap<unknown>(await response.json())
    if (result && typeof result === 'object') return result as Record<string, unknown>
    return { id: env.CLOUDFLARE_AI_SEARCH_INSTANCE }
  } catch (err) {
    if (err instanceof CloudflareAISearchError) throw err
    return { id: env.CLOUDFLARE_AI_SEARCH_INSTANCE, probe_error: String(err) }
  }
}

export async function uploadItem(
  env: Env,
  key: string,
  content: Uint8Array,
  contentType: string,
  timeoutMs = 60000,
): Promise<Record<string, unknown>> {
  requireConfigured(env)
  if (content.byteLength > MAX_FILE_BYTES) {
    throw new CloudflareAISearchError(
      `FILE_TOO_LARGE: Cloudflare AI Search rejects files over 4 MB (${content.byteLength} bytes)`,
    )
  }
  const form = new FormData()
  form.append('wait_for_completion', 'true')
  form.append('file', new Blob([content as BlobPart], { type: contentType }), key)
  const response = await fetch(instanceUrl(env, 'items'), {
    method: 'POST',
    headers: authHeaders(env),
    body: form,
    signal: AbortSignal.timeout(timeoutMs),
  })
  if (!response.ok) throw new CloudflareAISearchError(`Cloudflare upload failed (${response.status})`)
  const payload = (await response.json()) as Record<string, unknown>
  return unwrap<Record<string, unknown>>(payload)
}

export async function listItems(env: Env, timeoutMs = 60000): Promise<Record<string, unknown>[]> {
  requireConfigured(env)
  const items: Record<string, unknown>[] = []
  let page = 1
  for (;;) {
    const url = new URL(instanceUrl(env, 'items'))
    url.searchParams.set('page', String(page))
    url.searchParams.set('per_page', '50')
    url.searchParams.set('source', 'builtin')
    const response = await fetch(url.toString(), {
      headers: authHeaders(env),
      signal: AbortSignal.timeout(timeoutMs),
    })
    if (!response.ok) throw new CloudflareAISearchError(`Cloudflare list failed (${response.status})`)
    const payload = (await response.json()) as Record<string, unknown>
    const result = unwrap<unknown>(payload)
    const batch = asItemList(result)
    items.push(...batch)
    const info = (payload.result_info ?? {}) as Record<string, unknown>
    const total = info.total_count
    if (!batch.length || (typeof total === 'number' && items.length >= total)) break
    if (batch.length < 50) break
    page += 1
  }
  return items
}

export async function deleteItem(env: Env, itemId: string, timeoutMs = 60000): Promise<void> {
  requireConfigured(env)
  const response = await fetch(instanceUrl(env, `items/${encodeURIComponent(itemId)}`), {
    method: 'DELETE',
    headers: authHeaders(env),
    signal: AbortSignal.timeout(timeoutMs),
  })
  if (!response.ok) throw new CloudflareAISearchError(`Cloudflare delete failed (${response.status})`)
}

export interface SearchOptions {
  folderPrefix?: string
  maxNumResults?: number
  timeoutMs?: number
}

export async function search(
  env: Env,
  query: string,
  options: SearchOptions = {},
): Promise<Record<string, unknown>[]> {
  requireConfigured(env)
  const envTopK = Number.parseInt(env.RAG_RETRIEVAL_TOP_K ?? '8', 10)
  const fallback = Number.isFinite(envTopK) ? envTopK : 8
  const limit = Math.min(SEARCH_MAX_RESULTS, Math.max(1, options.maxNumResults ?? fallback))
  const body: Record<string, unknown> = {
    query: query || ' ',
    ai_search_options: {
      retrieval: { max_num_results: limit },
    },
  }
  if (options.folderPrefix) {
    ;(body.ai_search_options as Record<string, Record<string, unknown>>).retrieval.filters =
      folderStartsWithFilter(options.folderPrefix)
  }
  const response = await fetch(instanceUrl(env, 'search'), {
    method: 'POST',
    headers: { ...authHeaders(env), 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
    signal: AbortSignal.timeout(options.timeoutMs ?? 60000),
  })
  if (!response.ok) throw new CloudflareAISearchError(`Cloudflare search failed (${response.status})`)
  const payload = (await response.json()) as Record<string, unknown>
  const result = unwrap<unknown>(payload)
  if (Array.isArray(result)) return result.filter((c) => typeof c === 'object' && c !== null) as Record<string, unknown>[]
  if (result && typeof result === 'object') {
    const chunks = (result as Record<string, unknown>).chunks ?? (result as Record<string, unknown>).data
    if (Array.isArray(chunks)) return chunks.filter((c) => typeof c === 'object' && c !== null) as Record<string, unknown>[]
  }
  return []
}

export async function deleteItemsByPrefix(env: Env, prefix: string): Promise<number> {
  if (!isConfigured(env)) return 0
  const items = await listItems(env)
  const targets = items.filter((item) => {
    const key = String(item.key ?? '')
    return key.startsWith(prefix.endsWith('/') ? prefix : `${prefix}/`)
  })
  for (const item of targets) {
    const id = String(item.id ?? '')
    if (id) await deleteItem(env, id)
  }
  return targets.length
}

function asItemList(result: unknown): Record<string, unknown>[] {
  if (Array.isArray(result)) return result.filter((c) => typeof c === 'object' && c !== null) as Record<string, unknown>[]
  if (result && typeof result === 'object') {
    const r = result as Record<string, unknown>
    for (const key of ['items', 'data', 'objects']) {
      const nested = r[key]
      if (Array.isArray(nested)) return nested.filter((c) => typeof c === 'object' && c !== null) as Record<string, unknown>[]
    }
    if ('id' in r || 'key' in r) return [r]
  }
  return []
}